tests/test_core_schedule.py

```diff
@@ -1097,3 +1162,262 @@ def test_building_trips_dataframe(schedule):
         by=['route', 'trip', 'departure_time']).reset_index(drop=True)
 
     assert_frame_equal(df.sort_index(axis=1), correct_df.sort_index(axis=1))
+
+
+def test_generating_vehicles(schedule):
+    schedule.generate_vehicles()
+    assert_semantically_equal(schedule.vehicles, {'veh_3_bus': {'type': 'bus'}, 'veh_2_bus': {'type': 'bus'},
+                                         'veh_1_bus': {'type': 'bus'}, 'veh_4_bus': {'type': 'bus'}})
+
+
+def test_generating_vehicles_with_shared_vehicles_and_consistent_modes(mocker, schedule):
+    schedule.vehicles = {}
+    mocker.patch.object(DataFrame, 'drop',
+                        return_value=DataFrame({'vehicle_id': ['v_1', 'v_2', 'v_1', 'v_2', 'v_3'],
+                                                'type': ['bus', 'bus', 'bus', 'bus', 'rail']}))
+    schedule.generate_vehicles()
+    assert_semantically_equal(schedule.vehicles, {'v_1': {'type': 'bus'}, 'v_2': {'type': 'bus'},
+                                                  'v_3': {'type': 'rail'}})
+
+
+def test_generating_additional_vehicles_by_default(schedule):
+    r = Route(
+        route_short_name='N55',
+        mode='bus',
+        trips={'trip_id': ['some_trip_1'],
+               'trip_departure_time': ['16:23:00'],
+               'vehicle_id': ['some_bus_2']},
+        arrival_offsets=['00:00:00', '00:06:00'],
+        departure_offsets=['00:00:00', '00:06:00'],
+        id='new',
+        stops=[schedule.stop('1'),
+               schedule.stop('3')]
+    )
+    schedule.add_route('service', r)
+    # change existing vehicle types to be different from mode to test whether they are regenerated with default
+    # mode type
+    schedule.vehicles = {'veh_3_bus': {'type': '_bus'}, 'veh_4_bus': {'type': '_bus'}, 'veh_1_bus': {'type': '_bus'},
+                         'veh_2_bus': {'type': '_bus'}}
+    schedule.generate_vehicles()
+    assert_semantically_equal(schedule.vehicles, {'veh_3_bus': {'type': '_bus'}, 'veh_4_bus': {'type': '_bus'},
+                                         'veh_1_bus': {'type': '_bus'}, 'veh_2_bus': {'type': '_bus'},
+                                         'some_bus_2': {'type': 'bus'}})
+
+
+def test_generating_new_vehicles_with_overwite_True(schedule):
+    # change existing vehicle types to be different from mode to test whether they are regenerated with default
+    # mode type
+    schedule.vehicles = {'veh_3_bus': {'type': '_bus'}, 'veh_4_bus': {'type': '_bus'}, 'veh_1_bus': {'type': '_bus'},
+                         'veh_2_bus': {'type': '_bus'}}
+    schedule.generate_vehicles(overwrite=True)
+    assert_semantically_equal(schedule.vehicles, {'veh_3_bus': {'type': 'bus'}, 'veh_4_bus': {'type': 'bus'},
+                                         'veh_1_bus': {'type': 'bus'}, 'veh_2_bus': {'type': 'bus'}})
+
+
+def test_generating_vehicles_with_shared_vehicles_and_inconsistent_modes(mocker, schedule):
```

Comment:
A better test name might be something like `test_rejects_inconsistent_modes_when_generating_vehicles`

Test names like `does_something_when_something` can give a better clue about the intent of the test and the expected behaviour of the code under test.